### PR TITLE
Add Reference class to display string function

### LINKED ISSUE #1735: Initial and initial expression not working on dropdown items with the type 'reference'
**Describe the bug**
When using a dropdown item with the type 'reference', the initial value set always defaults to 'Not Answered'. In addition, when trying to select an option from the list, the selected option resets to 'Not Answered' as shown in the video below.

**Component**
SDC library

**To Reproduce**
N/A

**Expected behavior**
I expect the initial value set to be selected by default

**Screenshots**
When using a checkbox, the item works as expected as shown below:
https://user-images.githubusercontent.com/90775548/204476003-3d705ed1-f547-490f-bcb8-a2690f2bbaa5.mp4

However, when using a dropdown, the item does **not** work as expected as shown below:
https://user-images.githubusercontent.com/90775548/204475957-26bb4e50-4899-41a5-9328-5d7ba5bcc233.mp4

**Smartphone (please complete the following information):**
- Device: Samsung A12
- Android version: 11
- Build number: RP1A.200720.012.A127FXXU4AUK1
- Version: 30

**Additional context**
Here is the [questionnaire](https://github.com/google/android-fhir/files/10111167/sample-questionnaire.txt) used in the video above.

What led me to this issue is when I was trying to implement initial expression while using [answer expression](https://github.com/google/android-fhir/pull/1480).

**Would you like to work on the issue?**
Possibly the person who added support for the reference type can look into this

### PR DESCRIPTION
**Pull request by omarismail94**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1735 

**Description**
Add Reference type and obtain the field to display the string of on the UI.

Follow-up question: are there other class types we could be missing here other than `Reference`? 

**Alternative(s) considered**
N/A

**Type**
Choose one: (**Bug fix** | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

https://user-images.githubusercontent.com/44980219/204922839-c989e03e-540c-4014-bcf1-fe3de9a493d3.mp4

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Review by jingtang10 — APPROVED** [review]
thanks @omarismail94!

File: datacapture/src/test/java/com/google/android/fhir/datacapture/MoreQuestionnaireResponseItemAnswerComponentTest.kt [hunk 1]
```diff
@@ -47,6 +48,23 @@ import org.robolectric.annotation.Config
 class MoreQuestionnaireResponseItemAnswerComponentTest {
   private val context = ApplicationProvider.getApplicationContext<Application>()
 
+  @Test
+  fun `displayString() should return name value for reference`() {
+    val answer =
+      QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent()
+        .setValue(Reference().setDisplay("Newton"))
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/MoreQuestionnaireResponseItemAnswerComponentTest.kt:55]: actually do you need to consider the case where display is null for a reference? do you then need to display the actual reference value?

**Review by jingtang10 — CHANGES_REQUESTED** [review]
thanks @omarismail94!

**Comment by omarismail94** [comment]
@jingtang10  PTAL. I resolved your comment and also moved where the `displayString` function is called to `MoreAnswerOptions` to be in the same location where `displayString` value is called for `QuestionnaireItemAnswerOptionComponent`. 

I wanted to come up with one function to combine both like this:
```
internal val Questionnaire.QuestionnaireItemAnswerOptionComponent.displayString: String
  get() = displayString(null, value)

internal fun QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent.displayString(
  context: Context
) = displayString(context, value)


private fun displayString(context: Context?, value: Type?) : String =
  return when(value) {
   //logic
  }
    
```

however, the logic becomes divergent. An example: in `QuestionnaireItemAnswerOptionComponent`, if the `value` is `DateType`, we just return `value.primitiveValue()`; so if we pass 2022-01-01 it will return 2022-01-01.

For `DateType`s in  `QuestionnaireResponseItemAnswerComponent`, we return `value.localDate?.localizedString`; so if we pass 2022-01-01, it will return 01/01/22. 

Do we want to make what is displayed consistent, or is there a reason there is a difference between how `QuestionnaireItemAnswerOptionComponent`s  and ` `QuestionnaireResponseItemAnswerComponent`s are displayed?

**Comment by jingtang10** [comment]
> @jingtang10 PTAL. I resolved your comment and also moved where the `displayString` function is called to `MoreAnswerOptions` to be in the same location where `displayString` value is called for `QuestionnaireItemAnswerOptionComponent`.
> 
> I wanted to come up with one function to combine both like this:
> 
> ```
> internal val Questionnaire.QuestionnaireItemAnswerOptionComponent.displayString: String
>   get() = displayString(null, value)
> 
> internal fun QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent.displayString(
>   context: Context
> ) = displayString(context, value)
> 
> 
> private fun displayString(context: Context?, value: Type?) : String =
>   return when(value) {
>    //logic
>   }
>     
> ```
> 
> however, the logic becomes divergent. An example: in `QuestionnaireItemAnswerOptionComponent`, if the `value` is `DateType`, we just return `value.primitiveValue()`; so if we pass 2022-01-01 it will return 2022-01-01.
> 
> For `DateType`s in `QuestionnaireResponseItemAnswerComponent`, we return `value.localDate?.localizedString`; so if we pass 2022-01-01, it will return 01/01/22.
> 
> Do we want to make what is displayed consistent, or is there a reason there is a difference between how `QuestionnaireItemAnswerOptionComponent`s and ` `QuestionnaireResponseItemAnswerComponent`s are displayed?

I don't think they need to be different - please unify them.

**Comment by omarismail94** [comment]
@jingtang10 Unified the functions. With regards to existing tests that I needed to change, I only had to change [MoreAnswerOptionsTest](https://github.com/google/android-fhir/pull/1740/files#diff-6d2bb499344c625955cd158ed797e7047659a1ef898ba22daea16d57a1e18a4e); I had to change the display format of the Date

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt [hunk 2]
```diff
@@ -34,23 +46,61 @@ internal const val EXTENSION_OPTION_EXCLUSIVE_URL =
...
-      is Coding -> {
-        val display = valueCoding.displayElement.getLocalizedText() ?: valueCoding.display
-        if (display.isNullOrEmpty()) {
-          valueCoding.code
-        } else {
-          display
-        }
+  get() = displayString(null, value)
+
+internal fun QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent.displayString(
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt:51]: actually i think this should be in a separate file...

this file should be renamed to `MoreQuestionnaireItemAnswerOptionComponents.kt`.

and there should be another file `MoreQuestionnaireResponseItemAnswerComponents.kt`

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt (hunk 2 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt]: Created these two files, and third one called DisplayStringUtil and refactored the different functions to the right files

**Review by jingtang10 — COMMENTED** [review]
discussed with @omarismail94 - will refactor and make context mandatory, and then maybe move the extension function to `Type` instead.

**Comment by omarismail94** [comment]
> discussed with @omarismail94 - will refactor and make context mandatory, and then maybe move the extension function to `Type` instead.

implemented this

**Review by jingtang10 — APPROVED** [review]
thanks omar!

File: datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt [hunk 3]
```diff
@@ -28,3 +49,44 @@ fun Type.asStringValue(): String {
...
+      val display = this.displayElement.getLocalizedText() ?: this.display
+      if (display.isNullOrEmpty()) {
+        this.code ?: context.getString(R.string.not_answered)
+      } else display
+    }
+    is DateType -> this.localDate?.localizedString ?: context.getString(R.string.not_answered)
+    is DateTimeType ->
+      "${this.localDate.localizedString} ${this.localTime.toLocalizedString(context)}"
+    is DecimalType -> this.valueAsString ?: context.getString(R.string.not_answered)
+    is IntegerType -> this.valueAsString ?: context.getString(R.string.not_answered)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt:79]: can you merge these together?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt [hunk 4]
```diff
@@ -28,3 +49,44 @@ fun Type.asStringValue(): String {
...
+    is Reference -> {
+      val display = this.display
+      if (display.isNullOrEmpty()) {
+        this.reference ?: context.getString(R.string.not_answered)
+      } else display
+    }
+    is StringType -> this.getLocalizedText()
+        ?: this.valueAsString ?: context.getString(R.string.not_answered)
+    is TimeType -> this.valueAsString ?: context.getString(R.string.not_answered)
+    is UriType -> this.valueAsString ?: context.getString(R.string.not_answered)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt:90]: can you merge these as well?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt (hunk 3 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt (hunk 4 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/common/datatype/MoreTypes.kt]: Done

**Pull request merged**